draft-ietf-mls-protocol.md

```diff
@@ -814,11 +866,50 @@ struct {
   }
 } optional<T>;
 
+struct {
+  HPKEPublicKey public_key;
+  Credential credential;
+} LeafNodeInfo;
+
+struct {
+  uint8 hash_type = 0;
+  optional<LeafNodeInfo> info;
+} LeafNodeHashInput;
+~~~~~
+
+The `public_key` and `credential` fields represent the leaf public
+key and the credential for the member holding that leaf,
+respectively.  The `info` field is equal to the null optional value
+when the leaf is blank (i.e., no member occupies that leaf).
+
+Likewise, the hash of a parent node (including the root) is the hash
+of a `ParentNodeHashInput` struct:
+
+~~~~~
+struct {
+  uint8 hash_type = 1;
+  HPKEPublicKey public_key;
+  opaque left_hash<0..255>;
+  opaque right_hash<0..255>;
+} ParentNodeHashInput
+~~~~~
+
+The `left_hash` and `right_hash` fields hold the hashes of the
+node's left and right children, respectively.  The `public_key_hash`
```

Comment:
The text description doesn't seem to match the structs. 
 * The `public_key_hash` field is missing from the `ParentNodeHashInput` struct. To make it consistent with the text I think the `public_key` should be replaced with `opaque public_key_hash<0..255>;` or something like that. 
 * Also, for completeness, we might want to add something like the following struct as it's mentioned in the text (e.g. to make the `hash_type` field precise) :

`struct {
  uint8 hash_type = 2;
  optional <HPKEPublicKey> node_public_key;
} ParentNodeInfo`

More generally though, I have some comments/questions about this section:
* Why include `LeafNodeInfo` in `LeafNodeHashInput` directly but for parent nodes use *the hash of* `ParentNodeInfo` in `ParentNodeHashInput` instead (at least according to the text)? I don't see why the extra hash of NodeInfo makes sense for parents but not for leaves... Either both or neither no?
* What is the purpose of `hash_type`? If domain separation why is that needed and why is hash_type sufficient?
* On a related note, as defined now all blank leaves will have the same hash. Even internal nodes can have the same hash if they are both roots of identical sub-trees of blank nodes. I'm not sure its a problem... but if so one clean way to avoid this is to including `leaf_index` in the `LeafNodeHashInput` struct as a tie-breaker.

Reply:
My thinking with the pre-hashing of the parent nodes values is that if someone were providing incomplete views of the tree together with proofs, then the proof could just include the hash instead of the value.  But on further reflection, I think this is a mistake since (a) the only contents of the node are an `optional<HPKEPublicKey>` and (b) that's not meaningfully different from its hash.  So I've just eliminated the `_hash` part.

Re `hash_type` - TBH, I'm basically just copy/pasting from [Certificate Transparency](https://tools.ietf.org/html/rfc6962#section-2.1) here; open to arguments either way.  Maybe @beurdouche has some thoughts here?

Re blank nodes - Even more, a blank subtree anywhere in the tree will have the same value.  I had considered just throwing the node index into all the hashes for diversity.  But I don't really see an issue here -- why do we need to distinguish between blanks in different places? Given that in the ultimate root hash they'll end up tied with non-blank stuff and locked into their position that way.

Reply:
Thought a bit more about hash_type. For Merkle Trees (as in Certificate Transparency) the prefix to the hash inputs are definitely needed to separate leaf and internal node hashing domains. Otherwise MTs trivially aren't binding. E.g. given any target data set d={d(0),...,d(n)} it would be easy to make a new data point {d'} such that MT(d) = MT(d').

For MLS I'm suspect tree hashes may already be binding even without including hash_type. But justifying that would require subtle & messy reasoning including exactly how optional structs, HPKEPublicKey, SignaturesPK, Certificates, etc. are all serialized.

But that discussion can be entirely avoided by including hash_type as you have done. To see binding: hash_type ensures that for 2 tree's to have the same tree hash value they must have the same set of nodes and edges. (Otherwise two hash calls has differing hash_type inputs but results in the same output which is a collision on the underlying hash func). That in turn, means the trees must have the same set of labels for all nodes since otherwise we again have a collision somewhere. Ergo tree hash is binding.

Much cleaner. :-) So I've come round to the inclusion of hash_type even if I'm not convinced its strictly necessary.

Reply:
Thanks!  Belt and suspenders!